File: protocol/frame_v2.go
package protocol

import (
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

const (
	// Version2 enables stream multiplexing over a single WebSocket.
	Version2 uint8 = 2

	// HeaderSizeV2 is the v2 frame header size (includes StreamID).
	HeaderSizeV2 = 16

	TypeSession   uint8 = 7
	TypeSessionOK uint8 = 8
)

// EncodeTargetPayload serializes host+port for v2 CONNECT (password is session-scoped).
func EncodeTargetPayload(host string, port uint16) []byte {
	hostBytes := []byte(host)
	buf := make([]byte, 2+len(hostBytes)+2)
	off := putString(buf, hostBytes)
	binary.BigEndian.PutUint16(buf[off:], port)
	return buf
}

// DecodeTargetPayload deserializes v2 CONNECT payload.
func DecodeTargetPayload(data []byte) (host string, port uint16, err error) {
	off := 0
	host, n, err := getString(data, off)
	if err != nil {
		return "", 0, fmt.Errorf("decode target host: %w", err)
	}
	off += n
	if off+2 > len(data) {
		return "", 0, errors.New("decode target: missing port")
	}
	port = binary.BigEndian.Uint16(data[off:])
	return host, port, nil
}

// EncodeSessionPayload serializes a SESSION frame payload (password only).
func EncodeSessionPayload(password string) []byte {
	return putStringOnly([]byte(password))
}

// DecodeSessionPayload deserializes a SESSION frame payload.
func DecodeSessionPayload(data []byte) (string, error) {
	pass, _, err := getString(data, 0)
	return pass, err
}

func putStringOnly(s []byte) []byte {
	buf := make([]byte, 2+len(s))
	putString(buf, s)
	return buf
}

// EncodeFrameV2 serializes a v2 frame.
func EncodeFrameV2(f Frame) ([]byte, error) {
	if f.Version != Version2 {
		f.Version = Version2
	}
	if len(f.Payload) > MaxPayloadSize {
		return nil, ErrPayloadTooLarge
	}
	buf := make([]byte, HeaderSizeV2+len(f.Payload))
	binary.BigEndian.PutUint32(buf[0:4], Magic)
	buf[4] = Version2
	buf[5] = f.Type
	binary.BigEndian.PutUint32(buf[6:10], f.StreamID)
	binary.BigEndian.PutUint32(buf[10:14], uint32(len(f.Payload)))
	copy(buf[HeaderSizeV2:], f.Payload)
	return buf, nil
}

// DecodeFrameV2 reads a v2 frame from r.
func DecodeFrameV2(r io.Reader) (Frame, error) {
	header := make([]byte, HeaderSizeV2)
	if _, err := io.ReadFull(r, header); err != nil {
		return Frame{}, err
	}
	magic := binary.BigEndian.Uint32(header[0:4])
	if magic != Magic {
		return Frame{}, ErrInvalidMagic
	}
	if header[4] != Version2 {
		return Frame{}, ErrInvalidVersion
	}
	length := binary.BigEndian.Uint32(header[10:14])
	if length > MaxPayloadSize {
		return Frame{}, ErrPayloadTooLarge
	}
	payload := make([]byte, length)
	if length > 0 {
		if _, err := io.ReadFull(r, payload); err != nil {
			return Frame{}, err
		}
	}
	return Frame{
		Version:  Version2,
		Type:     header[5],
		StreamID: binary.BigEndian.Uint32(header[6:10]),
		Payload:  payload,
	}, nil
}

// DecodeFrameAny reads v1 or v2 based on the version byte after magic.
func DecodeFrameAny(data []byte) (Frame, error) {
	if len(data) < 5 {
		return Frame{}, errors.New("frame too short")
	}
	if binary.BigEndian.Uint32(data[0:4]) != Magic {
		return Frame{}, ErrInvalidMagic
	}
	switch data[4] {
	case Version:
		return DecodeFrame(newBytesReaderFrom(data))
	case Version2:
		return DecodeFrameV2(newBytesReaderFrom(data))
	default:
		return Frame{}, ErrInvalidVersion
	}
}

func newBytesReaderFrom(data []byte) *sliceReader {
	return &sliceReader{data: data}
}

type sliceReader struct {
	data []byte
	off  int
}

func (r *sliceReader) Read(p []byte) (int, error) {
	if r.off >= len(r.data) {
		return 0, io.EOF
	}
	n := copy(p, r.data[r.off:])
	r.off += n
	return n, nil
}

File: worker/src/protocol/frame_v2.ts
import {
  MAGIC,
  MAX_PAYLOAD_SIZE,
  VERSION,
  decodeFrame,
  type Frame,
  type FrameTypeValue,
} from "./frame";

export const VERSION2 = 2;
export const HEADER_SIZE_V2 = 16;

export const FrameTypeV2 = {
  SESSION: 7,
  SESSION_OK: 8,
} as const;

export function encodeTargetPayload(host: string, port: number): Uint8Array {
  const hostBytes = new TextEncoder().encode(host);
  const buf = new Uint8Array(2 + hostBytes.length + 2);
  const view = new DataView(buf.buffer);
  let off = putString(buf, view, 0, hostBytes);
  view.setUint16(off, port, false);
  return buf;
}

export function decodeTargetPayload(data: Uint8Array): { host: string; port: number } {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const [host, off1] = getString(data, view, 0);
  if (off1 + 2 > data.length) {
    throw new Error("target payload missing port");
  }
  const port = view.getUint16(off1, false);
  return { host, port };
}

export function encodeSessionPayload(password: string): Uint8Array {
  const passBytes = new TextEncoder().encode(password);
  const buf = new Uint8Array(2 + passBytes.length);
  putString(buf, new DataView(buf.buffer), 0, passBytes);
  return buf;
}

export function decodeSessionPayload(data: Uint8Array): string {
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const [password] = getString(data, view, 0);
  return password;
}

export function encodeFrameV2(frame: Frame & { streamId: number }): Uint8Array {
  if (frame.payload.length > MAX_PAYLOAD_SIZE) {
    throw new Error("payload too large");
  }
  const buf = new Uint8Array(HEADER_SIZE_V2 + frame.payload.length);
  const view = new DataView(buf.buffer);
  view.setUint32(0, MAGIC, false);
  buf[4] = VERSION2;
  buf[5] = frame.type;
  view.setUint32(6, frame.streamId, false);
  view.setUint32(10, frame.payload.length, false);
  buf.set(frame.payload, HEADER_SIZE_V2);
  return buf;
}

export function decodeFrameV2(data: Uint8Array): Frame & { streamId: number } {
  if (data.length < HEADER_SIZE_V2) {
    throw new Error("frame too short");
  }
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const length = view.getUint32(10, false);
  if (length > MAX_PAYLOAD_SIZE) {
    throw new Error("payload too large");
  }
  if (data.length < HEADER_SIZE_V2 + length) {
    throw new Error("incomplete frame");
  }
  return {
    version: VERSION2,
    type: data[5] as FrameTypeValue,
    flags: 0,
    streamId: view.getUint32(6, false),
    payload: data.slice(HEADER_SIZE_V2, HEADER_SIZE_V2 + length),
  };
}

export function decodeFrameAny(data: Uint8Array): Frame & { streamId: number } {
  if (data.length < 5) {
    throw new Error("frame too short");
  }
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  if (view.getUint32(0, false) !== MAGIC) {
    throw new Error("invalid magic");
  }
  const version = data[4];
  if (version === VERSION) {
    const frame = decodeFrame(data);
    return { ...frame, streamId: 1 };
  }
  if (version === VERSION2) {
    return decodeFrameV2(data);
  }
  throw new Error("unsupported version");
}

function putString(buf: Uint8Array, view: DataView, off: number, s: Uint8Array): number {
  view.setUint16(off, s.length, false);
  buf.set(s, off + 2);
  return off + 2 + s.length;
}

function getString(data: Uint8Array, view: DataView, off: number): [string, number] {
  if (off + 2 > data.length) {
    throw new Error("missing string length");
  }
  const length = view.getUint16(off, false);
  off += 2;
  if (off + length > data.length) {
    throw new Error("string exceeds payload");
  }
  const str = new TextDecoder().decode(data.slice(off, off + length));
  return [str, off + length];
}
